### Materials Project Workshop – July 31st to August 2nd, 2019, Berkeley, California

![pymatgen logo](http://pymatgen.org/_images/pymatgen.png)

# 0. What is pymatgen?

Pymatgen (Python Materials Genomics) is the code that powers all of the scientific analysis behind the Materials Project.  It includes a robust and efficient libraries for the handling of crystallographic structures and molecules, in addition to various mathematical and scientific tools for the handling and generation of materials data. For more details, see the [pymatgen website](https://pymatgen.org).

## 0.1 How do I install pymatgen?

For the workshop, pymatgen has been pre-installed for use in your Jupyter notebooks.

Otherwise, pymatgen can be installed via pip:

`pip install pymatgen`

or conda:

`conda install --channel matsci pymatgen`

We recommend using Python 3.6 or above. Until 2018, pymatgen was developed simultaneously for Python 2.x and 3.x, but following the rest of the Python community we have phased out support for Python 2.x, and since version 2019.1.1 we are developing exclusively for Python 3.x.

## 0.2 Where can I find help and how do I get involved?

* **For general help:** [pymatgen discourse](https://pymatgen.discourse.group/) is a place to ask questions.

* **To report bugs:** The [Github Issues](https://github.com/materialsproject/pymatgen/issues) page is a good place to report bugs.

* **For Materials Project data and website discussions:** The Materials Project  has its community [Materials Project Discussion](https://discuss.materialsproject.org) forum. 

* **For more example notebooks:** [matgenb](http://matgenb.materialsvirtuallab.org) is a new resource of Jupyter notebooks demonstrating various pymatgen functionality.

If you want specific new features, you're welcome to ask! We try to respond to community needs. If you're a developer and can add the feature yourself, we actively encourage you to do so by creating a Pull Request on Github with your additional functionality. To date, pymatgen has seen over 17,000 commits and over 120 contributors, and we try to have an inclusive and welcoming development community. All contributors are also individually acknowledged on [materialsproject.org/about](https://materialsproject.org/about).


# 1. Verify we have pymatgen installed

First, let's verify we have pymatgen installed. The following command should produce no error or warning:

In [1]:
import pymatgen

We can show the specific version of pymatgen installed:

In [2]:
print(pymatgen.__version__)

2019.7.21


For a list of new features, bug fixes and other changes, consult the [changelog on pymatgen.org](http://pymatgen.org/change_log.html).

You can also see where pymatgen is installed on your computer:

In [3]:
print(pymatgen.__file__)

/Users/samuelblau/pymatgen/pymatgen/__init__.py


We can also see which version of the Python programming language we are using:

In [4]:
import sys
print(sys.version)

3.7.3 (default, Mar 27 2019, 16:54:48) 
[Clang 4.0.1 (tags/RELEASE_401/final)]


If you have problems or need to report bugs when using pymatgen after the workshop, the above information is often very useful to help us identify the problem.

# 2. Structures and Molecules

Most of the fundamentals of pymatgen are expressed in terms of [**`Structure`**](http://pymatgen.org/pymatgen.core.structure.html#pymatgen.core.structure.Structure) and [**`Molecule`**](http://pymatgen.org/pymatgen.core.structure.html#pymatgen.core.structure.Molecule) objects.

While we will mostly be using `Structure`, `Stucture` and `Molecule` are very similar conceptually. The main difference is that `Structure` supports full periodicity required to describe crystallographic structures.

Creating a `Structure` can be done in one line, even for complicated crystallographic structures. However, let's start by introducing the simpler `Molecule` object, and then use this understanding of `Molecule` to introduce `Structure`.

## 2.0 Creating a Molecule

Start by importing `Molecule`:

In [5]:
from pymatgen import Molecule

In a Jupyter notebook, you can show help for any Python object by clicking on the object and pressing **Shift+Tab**. This will give you a list of arguments and keyword arguments necessary to construct the object, as well as the documentation ('docstring') which gives more information on what each argument means.

In [ ]:
Molecule

Molecule takes input **arguments** `species` and `coords`, and input **keyword arguments** `charge`, `spin_multiplicity`, `validate_proximity` and `site_properties`.

Keyword arguments come with a default value (the value after the equals sign), and so keyword arguments are optional.

Arguments (without default values) are mandatory.

In [7]:
my_molecule = Molecule(['C','O'], [[0, 0, 0], [0, 0, 1.2]])

In [8]:
print(my_molecule)

Full Formula (C1 O1)
Reduced Formula: CO
Charge = 0, Spin Mult = 1
Sites (2)
0 C     0.000000     0.000000     0.000000
1 O     0.000000     0.000000     1.200000


## 2.1 Visualizing a Molecule

It can be very helpful to quickly visualize molecules or structures. While this capability is still under development and not yet fully released, we will be using Crystal Toolkit's latest viewer during today's lesson.

In [9]:
from crystal_toolkit import view

In [10]:
view(my_molecule)

Renderer(background='white', camera=OrthographicCamera(bottom=-1.44, left=-1.44, near=-2000.0, position=(0.0, …

## 2.2 What's in a Molecule? Introducing Sites, Elements and Species

You can access properties of the molecule, such as the Cartesian coordinates of its sites:

In [20]:
print(my_molecule.cart_coords)

[[0.  0.  0. ]
 [0.  0.  1.2]]


or properties that are computed on-the-fly, such as its center of mass:

In [21]:
print(my_molecule.center_of_mass)

[0.         0.         0.68544132]


To see the full list of available properties and methods, press **Tab** after typing `my_molecule.` in your Jupyter notebook. There are methods used to modify the molecule and these take additional argument(s). For example, to add a charge to the molecule:

In [22]:
my_molecule.set_charge_and_spin(1)

In [23]:
print(my_molecule)

Full Formula (C1 O1)
Reduced Formula: CO
Charge = 1, Spin Mult = 2
Sites (2)
0 C     0.000000     0.000000     0.000000
1 O     0.000000     0.000000     1.200000


A molecule is essentially a list of `Site` objects. We can access these sites like we would a list in Python. For example, to obtain the total number of sites in the molecule:

In [24]:
len(my_molecule)

2

Or to access the first site (note that Python is a 0-indexed programming language, so the first site is site 0):

In [25]:
my_molecule[0]

Site: C (0.0000, 0.0000, 0.0000)

A site object contains information on the site's identity and position in space.

In [26]:
site0 = my_molecule[0]

In [27]:
site0.coords

array([0, 0, 0])

In [28]:
site0.specie

Element C

Here, the site holds the element C. In general, a site can hold either an [**`Element`**](http://pymatgen.org/pymatgen.core.periodic_table.html#pymatgen.core.periodic_table.Element), a [**`Specie`**](http://pymatgen.org/pymatgen.core.periodic_table.html#pymatgen.core.periodic_table.Specie) or a [**`Composition`**](http://pymatgen.org/pymatgen.core.composition.html#pymatgen.core.composition.Composition). Let's look at each of these in turn.

In [12]:
from pymatgen import Element, Specie, Composition

An `Element` is simply an element from the Periodic Table.

In [30]:
my_element = Element('C')

Elements have properties such as atomic mass, average ionic radius and more:

In [32]:
my_element.average_ionic_radius

0.3

A `Specie` can contain additional information, such as oxidation state:

In [33]:
Specie('O', oxidation_state=-2)

Specie O2-

Or, for convenience:

In [34]:
Specie.from_string('O2-')

Specie O2-

Finally, a `Composition` is an object that can hold certain amounts of different elements or species. This is most useful in a disordered Structure, and would rarely be used in a Molecule. For example, a site that holds 50% Au and 50% Cu would be set as follows:

In [35]:
Composition({'Au': 0.5, 'Cu': 0.5})

Comp: Cu0.5 Au0.5

When we construct a `Molecule`, the input argument will automatically be converted into one of `Element`, `Specie` or `Composition`. Thus, in the previous example, the input `['C', 'O']` was converted to `[Element C, Element O]`.

## 2.3 Exercise: Hydrogen Cyanide

Construct the linear HCN molecule where each bond distance is the sum of the two atomic radii. The atomic spheres should not substantially overlap.

In [13]:
H_rad = Element('H').atomic_radius
C_rad = Element('C').atomic_radius
N_rad = Element('N').atomic_radius
HC_bond_dist = H_rad + C_rad
CN_bond_dist = C_rad + N_rad
H_pos = 0
C_pos = H_pos + HC_bond_dist
N_pos = C_pos + CN_bond_dist
hcn = Molecule(['H','C','N'], [[0, 0, H_pos], [0, 0, C_pos],[0, 0, N_pos]])
view(hcn)

Renderer(background='white', camera=OrthographicCamera(bottom=-2.76, left=-2.76, near=-2000.0, position=(0.0, …

## 2.4 Creating a Structure and Lattice

Creating a `Structure` is very similar to creating a `Molecule`, except we now also have to specify a `Lattice`. 

In [14]:
from pymatgen import Structure, Lattice

A `Lattice` can be created in one of several ways, such as by inputting a 3x3 matrix describing the individual lattice vectors. For example, a cubic lattice of length 5 Ångstrom:

In [39]:
my_lattice = Lattice([[5, 0, 0], [0, 5, 0], [0, 0, 5]])

In [40]:
my_lattice

Lattice
    abc : 5.0 5.0 5.0
 angles : 90.0 90.0 90.0
 volume : 125.0
      A : 5.0 0.0 0.0
      B : 0.0 5.0 0.0
      C : 0.0 0.0 5.0

Equivalently, we can create it from its lattice parameters:

In [41]:
my_lattice_2 = Lattice.from_parameters(5, 5, 5, 90, 90, 90)

Or, since we know in this case that we have a cubic lattice, a == b == c and alpha == beta == gamma == 90 degrees, so we can simply put:

In [42]:
my_lattice_3 = Lattice.cubic(5)

We can confirm that these lattices are the same:

In [43]:
my_lattice == my_lattice_2 == my_lattice_3

True

Now, we can create a simple crystal structure. Let's start with body-centered-cubic iron:

In [15]:
bcc_fe = Structure(Lattice.cubic(2.8), ["Fe", "Fe"], [[0, 0, 0], [0.5, 0.5, 0.5]])

In [45]:
print(bcc_fe)

Full Formula (Fe2)
Reduced Formula: Fe
abc   :   2.800000   2.800000   2.800000
angles:  90.000000  90.000000  90.000000
Sites (2)
  #  SP      a    b    c
---  ----  ---  ---  ---
  0  Fe    0    0    0
  1  Fe    0.5  0.5  0.5


Creating this `Structure` was similar to creating a `Molecule` in that we provided a list of elements and a list of positions. However, there are two key differences: we had to include our `Lattice` object when creating the `Structure` and since we have a lattice, we can define the positions of our sites in *fractional coordinates* with respect to that lattice instead of Cartesian coordinates.

It's also possible to create an equivalent `Structure` using Cartesian coordinates:

In [46]:
bcc_fe_from_cart = Structure(Lattice.cubic(2.8), ["Fe", "Fe"], [[0, 0, 0], [1.4, 1.4, 1.4]],
                             coords_are_cartesian=True)

In [47]:
print(bcc_fe_from_cart)

Full Formula (Fe2)
Reduced Formula: Fe
abc   :   2.800000   2.800000   2.800000
angles:  90.000000  90.000000  90.000000
Sites (2)
  #  SP      a    b    c
---  ----  ---  ---  ---
  0  Fe    0    0    0
  1  Fe    0.5  0.5  0.5


We can check that both structures are equivalent:

In [48]:
bcc_fe == bcc_fe_from_cart

True

As in a `Molecule`, we can access properties of the structure, such as its volume:

In [49]:
bcc_fe.volume

21.951999999999995

And we can visualize the structure:

In [16]:
view(bcc_fe)

Renderer(background='white', camera=OrthographicCamera(bottom=-3.36, left=-3.36, near=-2000.0, position=(0.0, …

By default, the viewer only displays the lattice and the atoms explicitly defined in the structure. However, if we also want the periodic images of the atoms like you might be more used to, we can set *draw_image_atoms=True*:

In [17]:
view(bcc_fe,draw_image_atoms=True)

Renderer(background='white', camera=OrthographicCamera(bottom=-3.36, left=-3.36, near=-2000.0, position=(0.0, …

Note that the displayed bonds are inferred based on distance cutoffs.

## 2.5 Creating Structures from Spacegroups

Structures can also be created directly from their spacegroup:

In [44]:
bcc_fe = Structure.from_spacegroup("Im-3m", Lattice.cubic(2.8), ["Fe"], [[0, 0, 0]])
print(bcc_fe)

Full Formula (Fe2)
Reduced Formula: Fe
abc   :   2.800000   2.800000   2.800000
angles:  90.000000  90.000000  90.000000
Sites (2)
  #  SP      a    b    c
---  ----  ---  ---  ---
  0  Fe    0    0    0
  1  Fe    0.5  0.5  0.5


In [18]:
nacl = Structure.from_spacegroup("Fm-3m", Lattice.cubic(5.692), ["Na+", "Cl-"],
                                 [[0, 0, 0], [0.5, 0.5, 0.5]])
print(nacl)
view(nacl)

Full Formula (Na4 Cl4)
Reduced Formula: NaCl
abc   :   5.692000   5.692000   5.692000
angles:  90.000000  90.000000  90.000000
Sites (8)
  #  SP      a    b    c
---  ----  ---  ---  ---
  0  Na+   0    0    0
  1  Na+   0    0.5  0.5
  2  Na+   0.5  0    0.5
  3  Na+   0.5  0.5  0
  4  Cl-   0.5  0.5  0.5
  5  Cl-   0.5  0    0
  6  Cl-   0    0.5  0
  7  Cl-   0    0    0.5


Renderer(background='white', camera=OrthographicCamera(bottom=-6.8304, left=-6.8304, near=-2000.0, position=(0…

And spacegroups can be obtained from a structure:

In [44]:
nacl.get_space_group_info()

('Fm-3m', 225)

Where 225 is the spacegroup number.

## 2.6 Creating a Disordered Structure

Disordered structures are created using the syntax for compositions shown earlier. Here, we create a CuAu solid solution:

In [19]:
composition = {"Cu": 0.5, "Au":0.5}
cu_au = Structure.from_spacegroup("Fm-3m", Lattice.cubic(3.677), [composition], [[0, 0, 0]])
print(cu_au)
view(cu_au)

Full Formula (Cu2 Au2)
Reduced Formula: CuAu
abc   :   3.677000   3.677000   3.677000
angles:  90.000000  90.000000  90.000000
Sites (4)
  #  SP                    a    b    c
---  ------------------  ---  ---  ---
  0  Cu:0.500, Au:0.500  0    0    0
  1  Cu:0.500, Au:0.500  0    0.5  0.5
  2  Cu:0.500, Au:0.500  0.5  0    0.5
  3  Cu:0.500, Au:0.500  0.5  0.5  0


Renderer(background='white', camera=OrthographicCamera(bottom=-4.4124, left=-4.4124, near=-2000.0, position=(0…

## 2.7 Modifying a Structure

We can create a supercell by multiplying the structure by a number of repeats:

In [20]:
bcc_fe_repeated = bcc_fe*(2,2,2)

In [23]:
view(bcc_fe_repeated,draw_image_atoms=True)

Renderer(background='white', camera=OrthographicCamera(bottom=-6.72, left=-6.72, near=-2000.0, position=(0.0, …

We can remove a site:

In [24]:
nacl.remove_sites([0])
view(nacl)

Renderer(background='white', camera=OrthographicCamera(bottom=-6.8304, left=-6.8304, near=-2000.0, position=(0…

We can replace a site:

In [25]:
nacl.replace(0,'K')
view(nacl)

Renderer(background='white', camera=OrthographicCamera(bottom=-6.8304, left=-6.8304, near=-2000.0, position=(0…

We can translate a site:

In [26]:
nacl.translate_sites([0],[0,0.1,0])
view(nacl)

Renderer(background='white', camera=OrthographicCamera(bottom=-6.8304, left=-6.8304, near=-2000.0, position=(0…

We can also scale the lattice volume:

In [73]:
nacl.volume

184.41433388800002

In [64]:
nacl.scale_lattice(200)
nacl.volume

199.99999999999991

There are further *transformations* in pymatgen that can do more complicated structure manipulations, such as creating surfaces, grain boundaries or creating ordered approximations of disordered structures. Some of these will be discussed in the next lesson.

## 2.8 Exercise: N-V Center in Diamond

Build a diamond structure, which has a Fd-3m spacegroup, a cubic lattice of length 3.5 Angstrom, and a single carbon atom in its primitive cell. Then remove the carbon at the corner of the unit cell and replace its neighboring carbon with a nitrogen. Identify the new spacegroup.

In [27]:
diamond = Structure.from_spacegroup("Fd-3m", Lattice.cubic(3.5), ["C"],[[0, 0, 0]])
diamond.remove_sites([0])
diamond.replace(0,'N')
print(diamond.get_space_group_info())
view(diamond,draw_image_atoms=True)

('R3m', 160)


Renderer(background='white', camera=OrthographicCamera(bottom=-4.2, left=-4.2, near=-2000.0, position=(0.0, 0.…

The resulting nitrogen-vacancy center (N-V center) is a well-studied point defect in diamond which has photoluminescence that can be manipulated with magnetic, electric, or electromagnetic fields, yielding potential applications in spintronics and quantum computing.

# 3. Input and Output

## 3.0 Input/output for structures

Pymatgen can also read and write standard structure and molecule file formats such as:

  * [CIF](https://www.iucr.org/resources/cif)
  * [XCrySDen](http://www.xcrysden.org)
  * [xyz](https://en.wikipedia.org/wiki/XYZ_file_format)
  
while substantially more file formats are available using built-in interfaces with [Atomic Simulation Environment (ASE)](https://wiki.fysik.dtu.dk/ase/) and [Open Babel](http://openbabel.org/wiki/Main_Page).

For example, let's import from a CIF file:

In [28]:
struct = Structure.from_file('Nb2O5.cif')

In [29]:
print(struct)

Full Formula (Nb16 O40)
Reduced Formula: Nb2O5
abc   :   3.861885  14.714739  15.436378
angles: 104.895922  97.185452  97.526606
Sites (56)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  Nb    0.169637  0.247074  0.096428
  1  Nb    0.305803  0.511873  0.097557
  2  Nb    0.961103  0.023003  0.898803
  3  Nb    0.329079  0.328964  0.371122
  4  Nb    0.636261  0.67209   0.628918
  5  Nb    0.828076  0.753513  0.903124
  6  Nb    0.532266  0.408998  0.631902
  7  Nb    0.78282   0.939912  0.628586
  8  Nb    0.693758  0.488348  0.901824
  9  Nb    0.038548  0.978004  0.100936
 10  Nb    0.219435  0.060975  0.370982
 11  Nb    0.557784  0.216612  0.899574
 12  Nb    0.61544   0.862139  0.365606
 13  Nb    0.387379  0.138624  0.633853
 14  Nb    0.465408  0.592161  0.368071
 15  Nb    0.440698  0.783598  0.099606
 16  O     0.642207  0.232525  0.055067
 17  O     0.628144  0.344071  0.910055
 18  O     0.297975  0.092069  0.501416
 19  O     0.128816 

In [30]:
view(struct)

Renderer(background='white', camera=OrthographicCamera(bottom=-23.64246932739567, left=-23.64246932739567, nea…

Pymatgen also supports input writing and output parsing for a wide range of software packages:

* **Plane-wave DFT codes** including:
  * [VASP](https://www.vasp.at)
  * [Quantum ESPRESSO pwscf](https://www.quantum-espresso.org)
  * [ABINIT](https://www.abinit.org)
  * [exciting](http://exciting-code.org)
* **Quantum chemistry codes** including:
  * [Q-Chem](http://www.q-chem.com)
  * [Gaussian](http://gaussian.com)
  * [NWChem](http://www.nwchem-sw.org/index.php/Main_Page)


* **Many others, including ...**
    * [AiiDA](http://www.aiida.net)
    * [FEFF](http://feff.phys.washington.edu)
    * [ADF](https://www.scm.com/doc/ADF/index.html)
    * [LAMMPS](https://lammps.sandia.gov)
    * [Zeo++](http://www.zeoplusplus.org)
    * [Fiesta](http://perso.neel.cnrs.fr/xavier.blase/fiesta/)
    * [Phonopy](https://atztogo.github.io/phonopy/)
    * [ATAT (mcsqs)](https://www.brown.edu/Departments/Engineering/Labs/avdw/atat/)
    * [LOBSTER](http://www.cohp.de)

Further documentation can be found at [pymatgen.io](https://pymatgen.org/pymatgen.io.html).

There is also a lot of additional functionality in pymatgen for several of these software packages, including automating the generation of sensible input *sets* (configuration files) e.g. for VASP and Q-Chem.

# 3.1 Input/output within Materials Project codes

Most objects like `Structure` in Materials Project codes, including pymatgen, [atomate](https://atomate.org), [custodian](https://pythonhosted.org/custodian/), and [FireWorks](https://materialsproject.github.io/fireworks/), are "MSONable", named after "Monty JSON" from the [monty](http://guide.materialsvirtuallab.org/monty/) package. This means they can be easily converted to and from the JSON file format. This makes it really easy for us to pass objects between the different codes and also to store them in a database like [MongoDB](https://www.mongodb.com) or save them to disk.

Generally, objects that are MSONable will have `.as_dict()` and `.from_dict()` methods. To save or load from disk, the helpful `dumpfn` and `loadfn` functions from Monty can be used. These allow you to save both single Materials Project objects, e.g. a `Structure` object, as well as a list or dictionary of many kinds of objects in Materials Project codes.

In [31]:
from monty.serialization import dumpfn, loadfn

In [32]:
dumpfn(bcc_fe, "bcc_fe.json")
new_struct = loadfn("bcc_fe.json")
bandstructure = loadfn("li2o_bs.json")

In [33]:
bcc_fe.as_dict()

{'@module': 'pymatgen.core.structure',
 '@class': 'Structure',
 'charge': None,
 'lattice': {'matrix': [[2.8, 0.0, 0.0], [0.0, 2.8, 0.0], [0.0, 0.0, 2.8]],
  'a': 2.8,
  'b': 2.8,
  'c': 2.8,
  'alpha': 90.0,
  'beta': 90.0,
  'gamma': 90.0,
  'volume': 21.951999999999995},
 'sites': [{'species': [{'element': 'Fe', 'occu': 1}],
   'abc': [0.0, 0.0, 0.0],
   'xyz': [0.0, 0.0, 0.0],
   'label': 'Fe',
   'properties': {}},
  {'species': [{'element': 'Fe', 'occu': 1}],
   'abc': [0.5, 0.5, 0.5],
   'xyz': [1.4, 1.4, 1.4],
   'label': 'Fe',
   'properties': {}}]}

In [34]:
bandstructure.as_dict()

{'@module': 'pymatgen.electronic_structure.bandstructure',
 '@class': 'BandStructureSymmLine',
 'lattice_rec': {'@module': 'pymatgen.core.lattice',
  '@class': 'Lattice',
  'matrix': [[-1.350145435796662, 1.350145435796662, 1.350145435796662],
   [1.350145435796662, -1.350145435796662, 1.350145435796662],
   [1.350145435796662, 1.350145435796662, -1.350145435796662]]},
 'efermi': -0.0005,
 'kpoints': [[0.0, 0.0, 0.0],
  [0.01785714, 0.0, 0.01785714],
  [0.03571429, 0.0, 0.03571429],
  [0.05357143, 0.0, 0.05357143],
  [0.07142857, 0.0, 0.07142857],
  [0.08928571, 0.0, 0.08928571],
  [0.10714286, 0.0, 0.10714286],
  [0.125, 0.0, 0.125],
  [0.14285714, 0.0, 0.14285714],
  [0.16071429, 0.0, 0.16071429],
  [0.17857143, 0.0, 0.17857143],
  [0.19642857, 0.0, 0.19642857],
  [0.21428571, 0.0, 0.21428571],
  [0.23214286, 0.0, 0.23214286],
  [0.25, 0.0, 0.25],
  [0.26785714, 0.0, 0.26785714],
  [0.28571429, 0.0, 0.28571429],
  [0.30357143, 0.0, 0.30357143],
  [0.32142857, 0.0, 0.32142857],
  [0.3

# 3.2 Exercise: Barium Titanate

Load BaTiO3 from its CIF. Replace the barium with a new element with atomic number equal to the the day of the month you were born (e.g. 1st = hydrogen, 31st = gallium). Then make a supercell of N unit cells in one cartesian direction of your choice where N is the integer of your birth month (e.g. January = 1, December = 12). Dump the resulting structure to a file, swap structures with the person next to you using e-mail or Zulip. Load your partner's structure, visualize it to determine their birth month and examine the `Structure` object to determine their birth day. Using *draw_image_atoms=True* is recommended for visualization.

In [35]:
BaTiO3=Structure.from_file('BaTiO3.cif')
BaTiO3.replace(0,'H')
BaTiO3=BaTiO3*(1,1,5)
dumpfn(BaTiO3,"BaTiO3_sam.json")
struct=loadfn("BaTiO3_sam.json")
view(struct,draw_image_atoms=True)

Renderer(background='white', camera=OrthographicCamera(bottom=-24.652722928725705, left=-24.652722928725705, n…

In [36]:
print(struct)

Full Formula (Ti5 H5 O15)
Reduced Formula: TiHO3
abc   :   4.077159   4.077159  20.385796
angles:  89.699022  89.699022  89.699022
Sites (25)
  #  SP           a         b         c
---  ----  --------  --------  --------
  0  H     0.497155  0.497155  0.099431
  1  H     0.497155  0.497155  0.299431
  2  H     0.497155  0.497155  0.499431
  3  H     0.497155  0.497155  0.699431
  4  H     0.497155  0.497155  0.899431
  5  Ti    0.982209  0.982209  0.196442
  6  Ti    0.982209  0.982209  0.396442
  7  Ti    0.982209  0.982209  0.596442
  8  Ti    0.982209  0.982209  0.796442
  9  Ti    0.982209  0.982209  0.996442
 10  O     0.524065  0.012736  0.002547
 11  O     0.524065  0.012736  0.202547
 12  O     0.524065  0.012736  0.402547
 13  O     0.524065  0.012736  0.602547
 14  O     0.524065  0.012736  0.802547
 15  O     0.012736  0.012736  0.104813
 16  O     0.012736  0.012736  0.304813
 17  O     0.012736  0.012736  0.504813
 18  O     0.012736  0.012736  0.704813
 19  O     0.01273

# Summary

This notebook is intended to provide a short introduction to some of the functionality of pymatgen. We've examined the building blocks of pymatgen: the `Structure` and `Molecule` objects, the `Lattice`, `Element`, `Specie` and `Composition` objects that they're made up of, a way to visualize `Structure` and `Molecule` objects, simple `Structure` transformations, and some basic input/output functionality.